@here/harp-datasource-protocol/lib/StringEncodedNumeral.ts

```diff
@@ -24,105 +24,118 @@ export enum StringEncodedNumeralType {
     type: StringEncodedNumeralType.Meters,
     size: 1,
-    regExp: /((?=\.\d|\d)(?:\d+)?(?:\.?\d*))m/,
-    decoder: (encodedValue: string) => {
-        return [Number(StringEncodedMeters.regExp.exec(encodedValue)![1])];
+    regExp: /^((?=\.\d|\d)(?:\d+)?(?:\.?\d*))m$/,
+    decoder: (encodedValue: string, target: number[]) => {
+        const match = StringEncodedMeters.regExp.exec(encodedValue);
+        return match ? (target[0] = Number(match[1])) !== undefined : false;
     }
 };
 const StringEncodedPixels: StringEncodedNumeralFormat = {
     type: StringEncodedNumeralType.Pixels,
     size: 1,
     mask: 1.0,
-    regExp: /((?=\.\d|\d)(?:\d+)?(?:\.?\d*))px/,
-    decoder: (encodedValue: string) => {
-        return [Number(StringEncodedPixels.regExp.exec(encodedValue)![1])];
+    regExp: /^((?=\.\d|\d)(?:\d+)?(?:\.?\d*))px$/,
+    decoder: (encodedValue: string, target: number[]) => {
+        const match = StringEncodedPixels.regExp.exec(encodedValue);
+        if (match === null) {
+            return false;
+        }
+        target[0] = Number(match[1]);
+        return true;
     }
 };
 const StringEncodedHex: StringEncodedNumeralFormat = {
     type: StringEncodedNumeralType.Hex,
     size: 3,
-    regExp: /\#([0-9A-Fa-f]{1,8})/,
-    decoder: (encodedValue: string) => {
-        const match = StringEncodedHex.regExp.exec(encodedValue)!;
+    regExp: /^\#((?:[0-9A-Fa-f][0-9A-Fa-f]){3,4}|[0-9A-Fa-f]{3,4})$/,
+    decoder: (encodedValue: string, target: number[]) => {
+        const match = StringEncodedHex.regExp.exec(encodedValue);
+        if (match === null) {
+            return false;
+        }
         const hex = match[1];
         const size = hex.length;
         // Note that we simply ignore alpha channel value.
         // TODO: To be resolved with HARP-7517
         if (size === 3 || size === 4) {
             // #RGB or #RGBA
-            tmpColor.setRGB(
-                parseInt(hex.charAt(0) + hex.charAt(0), 16) / 255,
-                parseInt(hex.charAt(1) + hex.charAt(1), 16) / 255,
-                parseInt(hex.charAt(2) + hex.charAt(2), 16) / 255
-            );
+            target[0] = parseInt(hex.charAt(0) + hex.charAt(0), 16) / 255;
+            target[1] = parseInt(hex.charAt(1) + hex.charAt(1), 16) / 255;
+            target[2] = parseInt(hex.charAt(2) + hex.charAt(2), 16) / 255;
         } else if (size === 6 || size === 8) {
             // #RRGGBB or #RRGGBBAA
-            tmpColor.setRGB(
-                parseInt(hex.charAt(0) + hex.charAt(1), 16) / 255,
-                parseInt(hex.charAt(2) + hex.charAt(3), 16) / 255,
-                parseInt(hex.charAt(4) + hex.charAt(5), 16) / 255
-            );
+            target[0] = parseInt(hex.charAt(0) + hex.charAt(1), 16) / 255;
+            target[1] = parseInt(hex.charAt(2) + hex.charAt(3), 16) / 255;
+            target[2] = parseInt(hex.charAt(4) + hex.charAt(5), 16) / 255;
         } else {
+            // Impossible for given reg-exp
```

Comment:
Why not returning false instead of throwing an exception?

Reply:
I changed this to assert, cause it should never happen given that we have proper RegExp, we may also remove it entirely and then rely merely on unit tests.